### ISSUE TITLE
Exception when applying a baked migration snapshot

### ISSUE THREAD

**Issue by markstory** [cakephp/migrations#62]
_From @mamchenkov on March 23, 2015 10:29_

Steps to reproduce
-------------------------

```
# Create a new CakePHP 3 project
composer create-project --prefer-dist cakephp/app app
cd app
mysql -u root -e 'create database cakephp_test'
# Update db details
vim config/app.php
# Create table with id column set to binary(36) for UUID use
mysql -u root cakephp_test -e 'create table foo( id binary(36) not null unique, created datetime, modified datetime, primary key (id))'
# Bake snapshot
./bin/cake bake migration_snapshot initial
# Drop table
mysql -u root cakephp_test -e 'drop table foo'
# Apply snapshot
./bin/cake migrations migrate
```

Expected result
----------------------

Any baked migration should be usable.  In this case, a table should be correctly created and no exception should be triggered.

Actual result
-----------------

An exception is triggered, no table created:

```
$ ./bin/cake migrations migrate
Welcome to CakePHP v3.0.0 Console
---------------------------------------------------------------
App : src
Path: .../app/src/
---------------------------------------------------------------
using migration path .../app/config/Migrations
using environment default
using adapter mysql
using database cakephp_test

 == 20150323101149 Initial: migrating
                                                                                                                                
  [PDOException]                                                                                                                
  SQLSTATE[42000]: Syntax error or access violation: 1170 BLOB/TEXT column 'id' used in key specification without a key length 

migrate [-t|--target="..."] [-p|--plugin="..."] [-c|--connection="..."] [-s|--source="..."]
```

Additional Notes
---------------------
* If I use ```char(36)``` as column type, everything works as expected.
* The idea to use ```binary``` instead of ```char``` came up after [this StackOverflow thread](http://stackoverflow.com/questions/12646498/using-uuid-in-cakephp-what-datatype-is-recommended)

_Copied from original issue: cakephp/cakephp#6147_

**Comment by markstory** [comment]
_From @AD7six on March 23, 2015 10:39_

Please show the migration file.

What is the type of the id field _after_ the attempted migration? The error message says it's blob/text, not binary.

**Comment by markstory** [comment]
_From @mamchenkov on March 23, 2015 10:43_

@AD7six the migration file has ```binary```.  Here it is:

```
<?php
use Phinx\Migration\AbstractMigration;

class Initial extends AbstractMigration
{
    public function up()
    {
        $table = $this->table('foo', ['id' => false, 'primary_key' => ['id']]);
        $table
            ->addColumn('id', 'binary', [
                'default' => null,
                'limit' => 36,
                'null' => false,
            ])
            ->addColumn('created', 'datetime', [
                'default' => null,
                'limit' => null,
                'null' => true,
            ])
            ->addColumn('modified', 'datetime', [
                'default' => null,
                'limit' => null,
                'null' => true,
            ])
            ->create();
    }

    public function down()
    {
        $this->dropTable('foo');
    }
}
```

**Comment by markstory** [comment]
_From @mamchenkov on March 23, 2015 10:44_

There is no table at all after the migration.  I'm assuming that's because migration fails and Phinx rolls back the creation of the table.

**Comment by markstory** [comment]
_From @AD7six on March 23, 2015 10:58_

If I use that migration but a different field name (so it succeeds) I get this table:

```
CREATE TABLE `foo` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `bin` tinyblob NOT NULL, -- <- The primary key in the above migration
  `created` datetime DEFAULT NULL,
  `modified` datetime DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8
```

So it's not understood by phinx (the problem is the field type not "applying a baked migration snapshot")

**Comment by markstory** [comment]
_From @dakota on March 23, 2015 11:7_

Phinx currently only understands `char(36)` for uuid columns (If you set the column type to `uuid`). This sounds like a Phinx bug to me.

**Comment by markstory** [comment]
_From @mamchenkov on March 23, 2015 11:15_

Technically, yes, it's probably Phinx.  However, user-wise it's still CakePHP. :smiling_imp: .

**Comment by markstory** [comment]
_From @dereuromark on March 23, 2015 11:18_

If that doesn't work, sure sounds like a phinx issue then.

By the way:
Binary should definitely be the (documented) default over char(36) (it currently is still not, uuid will create char36.. http://book.cakephp.org/3.0/en/migrations.html#creating-custom-primary-keys ).
We should really change this and make this more clear IMO why binary would be the better way here.

**Comment by markstory** [comment]
_From @AD7six on March 23, 2015 11:20_

@mamchenkov sure - but for phinx/rob to fix it, it needs to know about the problem.

@dereuromark dunno about that, since we know for sure it currently won't work.

Related https://github.com/robmorgan/phinx/issues/85

**Comment by markstory** [comment]
_From @dereuromark on March 23, 2015 11:22_

Afterwards - as soon as we know it works (again) - I would really like to point out the more sane default value (regarding performance there is a huge difference).

Related: http://iops.io/blog/storing-billions-uuid-fields-mysql-innodb/ etc

**Comment by josegonzalez** [comment]
Phinx autogenerates an `id` field, which may be causing issues here.

**Comment by AD7six** [comment]
@josegonzalez  nah - the options override it:

    $table = $this->table('foo', ['id' => false, 'primary_key' => ['id']]);
                                  ^ No automatic autoinc primary key

**Comment by HavokInspiration** [comment]
I took a look at this.
Apparently, if you define a column as ``binary`` in phinx for MySQL, it is changed to a ``blob``, no matter what.

[See ref in Phinx source](https://github.com/robmorgan/phinx/blob/0.4.x-dev/src/Phinx/Db/Adapter/MysqlAdapter.php#L745-L761)

**Comment by HavokInspiration** [comment]
I proposed a fix on the phinx repo : https://github.com/robmorgan/phinx/pull/640

**Comment by HavokInspiration** [comment]
The issue was resolved with the robmorgan/phinx#640 PR that was merged.
The feature will be available with the next release of the plugin.
